### PR TITLE
[Trainer] GSPO support

### PR DESCRIPTION
**Pull request by bthecohen**
This PR adds support for [Group Sequence Policy Optimization (GSPO)](https://arxiv.org/abs/2507.18071), the hotness du jour from Alibaba Qwen. The implementation in this PR is loosely based on [this one](https://github.com/huggingface/trl/pull/3775) from TRL. It adds an `importance_sampling_level` config option which can be `token` (PPO/GRPO) or `sequence` (GSPO).

I ran a short/small GSM8k run with Qwen2.5-0.5B and the loss curves look okay:
image

However, I had to hack a few things to get this to run on Datadog's cloud infra (including changing some dependency versions) so I'd encourage one of the maintainers to reproduce these results locally before merging.

### CODE REVIEW

**Comment by bthecohen** [comment]
One thing a bit unexpected about this run was that the clip ratio was actually _lower_ for GSPO, which is the opposite of what the paper says:

image

A few hypotheses about why this could be, in roughly descending order of likelihood IMO:

- I didn't run the training long enough (in the paper, they report the clip ratios averaged over a full post-training run). It does seem like towards the end the ratio was going up for GSPO
- There's something different about the model / data / hyperparams in my setup that led to the opposite behavior of what the Qwen folks saw
- There's a bug in how we're calculating the clip ratio
- There's a bug in my GSPO implementation

@tyler-griggs I'd be curious to see if you see the same thing with a bigger model / longer run

**Comment by erictang000** [comment]
hi @bthecohen, thanks for the PR! I'll try testing this out and let you know what I see!

**Comment by erictang000** [comment]
btw @bthecohen what `eps_clip_low`/`eps_clip_high` values did you use for the runs? It seems like the comparison in the paper is between 0.2/0.27 for GRPO and 4e-4/3e-4 for GSPO (https://x.com/ChujieZheng/status/1948933507696525392), so maybe that explains the diff in clip ratio?

**Comment by bthecohen** [comment]
Ah yeah that's it. I left this as the default (so 0.2/0.2 for both runs).

**Comment by bthecohen** [comment]
One additional thing to note that I realized this morning is that this implementation isn't quite correct when you have token-level advantages - for that case I think we should be using "GSPO-token" (equations 13 and 14 in the paper). As long as the per-token advantages are all the same within a sequence, they are equivalent but given that SkyRL now supports returning advantages as a list per-token, I think we'll either want to:

- add explicit support for gspo-token with the stop-gradient correction
- give a warning that the current implementation shouldn't be used with per-token advantages

I didn't notice this at first because the TRL implementation doesn't support it. WDYT, is it worth blocking this PR on getting that working? Given the focus on multi-turn rollouts it might be important.

**Comment by erictang000** [comment]
yeah I think adding explicit support for gspo-token would be good given that the regular gspo case is just a subset of it where advantages are all identical - maybe we could do something similar to the structure of [this verl PR](https://github.com/volcengine/verl/pull/2775), and add the GSPO loss function as a separate policy loss function to avoid bloating the loss logic as we add more algorithm support?

I'll draft a quick PR to update the logic to make adding a new policy loss cleaner

**Comment by bthecohen** [comment]
I agree having modular policy loss functions is a good idea in general, but for the sake of argument we still might want the various PPO  flavors to share the same class...I'm just thinking people want to mix and match the GSPO stuff with the DAPO loss reduction, which could lead to some code duplication. An alternative approach would to be keep one class for PPO/GRPO/GSPO, but factor out some of the subcomputations into private methods to make the `forward` less bloated.

I don't feel super strongly about this – if you'd prefer it with separate classes I'm happy to rebase off of your PR. Let me know, I'm also happy to help with the refactor if I can. And revise this to use GSPO-token.

On a separate note, from that PR we have the lead author of GSPO saying that it might only matter for MoE models, or only in the SFT -> RL setting:

image

**Comment by bthecohen** [comment]
Actually, I guess verl solves this by just factoring out the loss reduction into a helper function. Although, for GSPO maybe only sequence-mean reduction makes sense?

**Comment by erictang000** [comment]
hey @bthecohen finished updating the logic for adding custom/new policy losses in #126, could you take a look and let me know if it makes sense to you?

once that's merged then gspo/gspo-token would just be a new policy loss implementation here (and we could reuse the loss reduction function with sequence mean hard coded)

**Comment by erictang000** [comment]
hey @bthecohen, now that #126 is merged, can we rebase this PR and add GSPO token? 

happy to help/run any longer test runs for it as well!

**Comment by bthecohen** [comment]
Yep, working on it now! Given the refactor, and the fact that I'm changing the algorithm, I may decide to close this PR and reopen rather than rebasing. But I'll see if I can keep it in here to keep the discussion thread linear.

**Comment by bthecohen** [comment]
@erictang000 I've updated this to use the new refactoring and to use GSPO token. New training results on my end:

image

The GSPO run here is using the clipping values from the paper. The GRPO run is still using the default hyperparams from the SkyRL example, so not exactly as in the paper.

**Comment by bthecohen** [comment]
The clip ratio is now much higher for GSPO, as expected 
image

**Comment by bthecohen** [comment]
Also, the eval curves look extremely close between GSPO and GRPO
image

**Comment by bthecohen** [comment]
Update: ran a longer GSPO run, looks quite stable 
image
